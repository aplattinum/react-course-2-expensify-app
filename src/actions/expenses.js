import uuid from 'uuid';

//---ACTIONS---//
//-------------//
//send the objects, arrays, or new data to the reducers to be processed there

//ADD_EXPENSE - adds a new object on to the array of objects
export const addExpense = (
    { description = '',
      note = '', 
      amount = 0, 
      createdAt= 0 
    } 
    = {}) => ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
    }
});

//REMOVE_EXPENSE - sends the id as string { id } --> pulls id out from expenses
export const removeExpense = ({ id }= {}) => ( { 
    type: 'REMOVE_EXPENSE',
    id
});

//Edit Expense - sends id object and updates object
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
});