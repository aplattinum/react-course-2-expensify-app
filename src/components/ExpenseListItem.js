//Export stateless functional component
//render description, amount, created at
import React from 'react';

import { Link } from 'react-router-dom';

const ExpenseListItem = ( {description, amount, createdAt, id, dispatch}) => (
    <div>
        <Link
        to={`/edit/${id}`}
        
        ><h3>{description}</h3></Link>
        <p>Amount: {amount} - Created On:{createdAt}</p>
        
    </div>
)

// const mapStateToProps = (state) => {
//     return {
//         expenses: state.expenses
//     }
// }

export default ExpenseListItem

