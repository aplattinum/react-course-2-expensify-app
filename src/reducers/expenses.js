
//Reducer recieves the action and objects (or arrays) from the action
//generator and defines what happens to the state object at the moment
// of being called.  
//EXPENSES REDUCER
const expensesReducerDefaultState = [];

export default (state= expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            return [
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            //filters out the selected id to be removed.
            //any id that isnt the selected id to be removed will 
            //pass through the filter.
            return state.filter(({ id }) => id !== action.id)
        case 'EDIT_EXPENSE':
            //finds the expense with the matching id.  Returns a new
            //object by spreading updates over the expense object -->
            //Rewriting any previous data and adding new data.
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {
                        ...expense, 
                        ...action.updates
                    }
                } else {
                    return expense
                }
            })
        case 'SET_EXPENSES':
            return action.expenses
            
        default:
          return state;
    }
}

