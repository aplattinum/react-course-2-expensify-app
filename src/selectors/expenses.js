import moment from 'moment';
//Get visible expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {

        const createdAtMoment = moment(expense.createdAt)
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day") : true; 
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, "day") : true;
        const textMatch =  expense.description.toLowerCase().includes( text.toLowerCase() )
        //Each of the const above must evaluate to true in order for the data to pass through the filter
        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        //newest to oldest 
        if (sortBy === 'date') {
            //return -1
            return a.createdAt < b.createdAt ? 1 : -1
        } else if (sortBy === 'amount') {
            return a.amount < b.amount ? 1 : -1
        }

    });
};
