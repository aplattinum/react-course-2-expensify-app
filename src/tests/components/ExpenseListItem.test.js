import ExpenseListItem from '../../components/ExpenseListItem';
import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';

test('Render expenseListItem with fixtures data', () => {
    const wrapper = shallow(<ExpenseListItem { ...expenses[0] }/>)
    expect(wrapper).toMatchSnapshot();
})
