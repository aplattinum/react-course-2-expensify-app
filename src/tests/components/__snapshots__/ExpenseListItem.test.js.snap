// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render expenseListItem with fixtures data 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h3>
      gum
    </h3>
  </Link>
  <p>
    Amount: 
    195
     - Created On:
    0
  </p>
</div>
`;
